sktime/classification/early_classification/_teaser.py

```diff
@@ -0,0 +1,429 @@
+            # A List containing the state info for case, edited at each time stamp.
+            # contains 1. the index of the time stamp, 2. the number of consecutive
+            # positive decisions made, and 3. the prediction made
+            state_info = [(0, 0, 0) for _ in range(n_instances)]
+            # Stores whether we have made a final decision on a prediction, if true
+            # state info wont be edited in later time stamps
+            finished = [False for _ in range(n_instances)]
+
+            for i in range(len(self._classification_points)):
+                if i == len(self._classification_points) - 1:
+                    decisions = [True for _ in range(n_instances)]
+                elif self._svms[i] is not None:
+                    decisions = self._svms[i].predict(X_svm[i]) == 1
+                else:
+                    decisions = [False for _ in range(n_instances)]
+
+                # record consecutive class decisions
+                state_info = [
+                    (
+                        # the classification point index
+                        i,
+                        # consecutive predictions, add one if positive decision and same
+                        # class
+                        state_info[n][1] + 1
+                        if decisions[n] and train_preds[i][n] == state_info[i][2]
+                        # set to 0 if the decision is negative, 1 if its positive but
+                        # different class
+                        else 1 if decisions[n] else 0,
+                        # predicted class index
+                        train_preds[i][n],
+                    )
+                    # if we have finished with this case do not edit the state info
+                    if not finished[n] else state_info[n]
+                    for n in range(n_instances)
+                ]
+
+                # safety decisions
+                finished = [
+                    True if state_info[n][1] >= g else False for n in range(n_instances)
+                ]
+
+            # calculate harmonic mean from finished state info
+            accuracy = (
+                np.sum(
+                    [
+                        1 if state_info[i][2] == self._class_dictionary[y[i]] else 0
+                        for i in range(n_instances)
+                    ]
+                )
+                / n_instances
+            )
+            earliness = (
+                1
+                - np.sum(
+                    [
+                        self.classification_points[state_info[i][0]] / series_length
+                        for i in range(n_instances)
+                    ]
+                )
+                / n_instances
+            )
+            hm = (2 * accuracy * earliness) / (accuracy + earliness)
+
+            if hm > best_hm:
+                best_hm = hm
+                self._consecutive_predictions = g
+
+        return self
+
+    def _predict(self, X):
```

Comment:
Not sure: what is the expected output of _predict() in early classification? 

As I understand it right now, it is a list of  probabilities for each offset? Not the overall decision after the full length?

I guess, I would rather model it as the overall decision. And have another method, to return the array of decisions.

Reply:
Unfortunately, we are sort of constrained by the sktime classifier interface unless we want to create new stuff specifically for early classification. It definitely could be better, but would require a new base class and set of tests most likely.

_predict() is required and has to return a list of class predictions, currently the only different for early classifiers is that it can return predictions for different series lengths.

_predict_proba() returns a list of probabilties for the input, assuming it is a valid input length. We leave the decision on whether those are safe or not to decide_prediction_safety().

Reply:
Hmm, still the contract of _predict_proba() is not clear to me. 

`idx = self._classification_point_dictionary.get(series_length, -1)`
I am not sure, if I understand the code correctly. It reads to me like `_predict_proba` takes the full length of the time series and returns the probability for that full length - am I correct about that?


I would expect _predict_proba() to probe the time series at the known offsets and return the probability of the time stamp where the earliest decision taken out. And there should be some method to get this time stamp.

Reply:
If your time stamps are [6, 10, 16, 24], _predict_proba() will accept series of those lengths then return the probability from the corresponding estimator. `idx = self._classification_point_dictionary.get(series_length, -1)` just checks which of [6, 10, 16, 24] the input series length is, or returns -1 if it's none of them so an error can be thrown.

We can't edit _predict_proba() to return null/a decision because of the classifier interface/unit tests I'm pretty sure. That's why it's done as a separate method the user has to call.

The problem with checking all the offsets at once is it requires the full series, and if you require the full series you can't really do early classification in practice.

Reply:
I have a different understanding of `_predict_proba()`:

Assuming we have given
(a) time stamps: [6, 10, 16, 24]  and
(b) time series has length 30. 

Then we probe all offsets 6, 10, 16, 24 and return the **first index** where the decision would be taken out. I.e. 3 consecutive positive votes.

If no decision is taken out earlier than the last index (24), we could 
(a) return the probability of the classifier at the last index (ndex=24) or 
(b) simply 0 for all classes.

For that you do not need to return null.

Reply:
Are we assuming 30 is the final length of the series? Can you also input a set of series of length 16 and also get probabilities?

If yes: You are doing a lot of repeated work if you want to attempt to make a decision at each successive time point.
If no: You require the full series to make predictions, the classifier is only useful for research purposes.

I may be misunderstanding a bit.

Reply:
I don't think the current set-up is the end-state we should be aiming for, having it all encapsulated into predict/predict_proba would be best. That would require a little extra time/drive though to set up interfaces and tests. Will add some bits to #1616.

Reply:
> If yes
Yes :)

> You are doing a lot of repeated work if you want to attempt to make a decision at each successive time point.
Not sure what you mean by that. Yes, if you are assuming a streaming setting, extending the time series and doing predict() after each new bit of information, you will have extra work.

But, there is no point in checking for consecutive votes in fit(), if we only take the last index in predict()...
Thus, imo, there have to be consecutive votes in predict(), too...

Reply:
The way it is currently set up, the decision-making process is separate from the probability prediction process. All predict/predict_proba does essentially is find the correct WEASEL and make predictions.

`decide_prediction_safety(X, X_probabilities, state_info)` Takes X, the data with a series length contained in [6, 10, 16, 24, 30]. X_probabilities, the result of calling predict_proba() on X. state_info, decision/prediction information from the previous decision made (or None if it's the first/6).

We only have to check the current index, as we save relevant information (the number of consecutive predictions, the last class predicted) in state_info. It is assumed that if you are trying to make a decision on length 16, you have already run it for length 10 and have the state_info.

We probably could put this all in predict_proba, and ideally we would, but currently that would mean we have to repeat predictions of all previous data points and figure out how to return negative decisions (I don't know if we have a test to make sure probabilities sum to 1, for the all 0 suggestion).

Reply:
so i know i just wrote an essay above, but I'm pretty sure I could make it so that predict_proba() takes an optional state_info i.e. `_predict_proba(X, state_info=None)` and have it only return decisions i.e. (probas, decisions) if a state_info was given or the series length is < the max. Pretty sure that would pass all the tests

Reply:
would have to write a bit for running on all previous time stamps, however.

Reply:
> We only have to check the current index, as we save relevant information (the number of consecutive predictions, the last class predicted) in state_info. It is assumed that if you are trying to make a decision on length 16, you have already run it for length 10 and have the state_info.

Sorry again - and glad to help rubber-ducking :)

I do not quite understand the implication of this. `predict()` can be called from outside at any time point by the user.

Thus it would be great to start this process probing all offsets, if there is no `state_info` available, and continue with the next available time stamp if state_info is available :)

Reply:
Ok, I have redone this and it is all integrated into predict_proba. It is not completely tested yet, will do that tomorrow.

If no state_info is provided, it predicts all time stamps up to the current series_length. If a state info is provided, it uses it and predicts for the current series length only.

Should be noted, it is possible for these to have different outputs currently. Case 1 will stop updating when a decision is made, Case 2 may update its decision even if it considered it safe in a previous call.

Reply:
Thank you. I will check the code. I think, passing one `state_info=state_info` was missing, which I fixed. 

I would like to go over the code. Still trying to understand it :)

Reply:
I have refactored the code